In [1]:
import numpy as np
import pandas as pd
import sklearn as sk
import matplotlib
import matplotlib.pyplot as plt

In [2]:
df_agency = pd.read_csv("AgencyDataset.csv")
df_agency.head(5)

AGENCY_ID  PRIMARY_AGENCY_ID   PROD_ABBR PROD_LINE STATE_ABBR  \
0        3.0                  3  BOILERMACH        CL         IN   
1        3.0                  3  BOILERMACH        CL         IN   
2        3.0                  3  BOILERMACH        CL         IN   
3        3.0                  3  BOILERMACH        CL         IN   
4        3.0                  3  BOILERMACH        CL         IN   

   STAT_PROFILE_DATE_YEAR  RETENTION_POLY_QTY  POLY_INFORCE_QTY  \
0                  2005.0                 0.0               0.0   
1                  2006.0                 0.0               0.0   
2                  2007.0                 0.0               0.0   
3                  2008.0                 0.0               0.0   
4                  2009.0                 0.0               0.0   

   PREV_POLY_INFORCE_QTY  NB_WRTN_PREM_AMT          ...            \
0                    0.0              40.0          ...             
1                    0.0             151.0          ...             
2                    0.0              40.0          ...             
3                    0.0              69.0          ...             
4                    0.0              28.0          ...             

   PL_BOUND_CT_ELINKS  PL_QUO_CT_ELINKS  PL_BOUND_CT_PLRANK  PL_QUO_CT_PLRANK  \
0                 0.0               0.0                 0.0             103.0   
1                 0.0               0.0                 0.0             103.0   
2                 0.0               0.0                 0.0             103.0   
3                 0.0               0.0                 0.0             103.0   
4                 0.0               0.0                 0.0             103.0   

   PL_BOUND_CT_eQTte  PL_QUO_CT_eQTte  PL_BOUND_CT_APPLIED  PL_QUO_CT_APPLIED  \
0               50.0            288.0                  0.0                0.0   
1               50.0            288.0                  0.0                0.0   
2               50.0            288.0                  0.0                0.0   
3               50.0            288.0                  0.0                0.0   
4               50.0            288.0                  0.0                0.0   

   PL_BOUND_CT_TRANSACTNOW  PL_QUO_CT_TRANSACTNOW  
0                      0.0                    0.0  
1                      0.0                    0.0  
2                      0.0                    0.0  
3                      0.0                    0.0  
4                      0.0                    0.0  

[5 rows x 49 columns]

In [3]:
df_agency.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190571 entries, 0 to 190570
Data columns (total 49 columns):
AGENCY_ID                    190569 non-null float64
PRIMARY_AGENCY_ID            190571 non-null int64
PROD_ABBR                    190570 non-null object
PROD_LINE                    190570 non-null object
STATE_ABBR                   190570 non-null object
STAT_PROFILE_DATE_YEAR       190570 non-null float64
RETENTION_POLY_QTY           190570 non-null float64
POLY_INFORCE_QTY             190570 non-null float64
PREV_POLY_INFORCE_QTY        190570 non-null float64
NB_WRTN_PREM_AMT             190570 non-null float64
WRTN_PREM_AMT                190570 non-null float64
PREV_WRTN_PREM_AMT           190570 non-null float64
PRD_ERND_PREM_AMT            190570 non-null float64
PRD_INCRD_LOSSES_AMT         190570 non-null float64
MONTHS                       190570 non-null float64
RETENTION_RATIO              190570 non-null float64
LOSS_RATIO                   190570 non-null f

In [4]:
df_bart = df_agency[["AGENCY_ID", "PRIMARY_AGENCY_ID", "STAT_PROFILE_DATE_YEAR", "MONTHS", "PRD_INCRD_LOSSES_AMT", "RETENTION_RATIO", "LOSS_RATIO","LOSS_RATIO_3YR", "GROWTH_RATE_3YR", "CL_BOUND_CT_MDS", "CL_QUO_CT_MDS","CL_BOUND_CT_SBZ", "CL_QUO_CT_SBZ","CL_BOUND_CT_eQT", "CL_QUO_CT_eQT","PL_BOUND_CT_ELINKS", "PL_QUO_CT_ELINKS","PL_BOUND_CT_PLRANK", "PL_QUO_CT_PLRANK","PL_BOUND_CT_eQTte", "PL_QUO_CT_eQTte","PL_BOUND_CT_APPLIED", "PL_QUO_CT_APPLIED","PL_BOUND_CT_TRANSACTNOW", "PL_QUO_CT_TRANSACTNOW"]]

In [5]:
df_bart.head(100)

AGENCY_ID  PRIMARY_AGENCY_ID  STAT_PROFILE_DATE_YEAR  MONTHS  \
0         3.0                  3                  2005.0     8.0   
1         3.0                  3                  2006.0    12.0   
2         3.0                  3                  2007.0    12.0   
3         3.0                  3                  2008.0    12.0   
4         3.0                  3                  2009.0    12.0   
5         3.0                  3                  2010.0    12.0   
6         3.0                  3                  2011.0    12.0   
7         3.0                  3                  2012.0    12.0   
8         3.0                  3                  2013.0    12.0   
9         3.0                  3                  2014.0    12.0   
10        3.0                  3                  2015.0     5.0   
11        3.0                  3                  2005.0     8.0   
12        3.0                  3                  2006.0    12.0   
13        3.0                  3                  2007.0    12.0   
14        3.0                  3                  2008.0    12.0   
15        3.0                  3                  2009.0    12.0   
16        3.0                  3                  2010.0    12.0   
17        3.0                  3                  2011.0    12.0   
18        3.0                  3                  2012.0    12.0   
19        3.0                  3                  2013.0    12.0   
20        3.0                  3                  2014.0    12.0   
21        3.0                  3                  2015.0     5.0   
22        3.0                  3                  2005.0     8.0   
23        3.0                  3                  2006.0    12.0   
24        3.0                  3                  2007.0    12.0   
25        3.0                  3                  2008.0    12.0   
26        3.0                  3                  2009.0    12.0   
27        3.0                  3                  2010.0    12.0   
28        3.0                  3                  2011.0    12.0   
29        3.0                  3                  2012.0    12.0   
..        ...                ...                     ...     ...   
70        3.0                  3                  2014.0    12.0   
71        3.0                  3                  2015.0     5.0   
72        3.0                  3                  2005.0     3.0   
73        3.0                  3                  2006.0    12.0   
74        3.0                  3                  2007.0    12.0   
75        3.0                  3                  2008.0     2.0   
76        3.0                  3                  2009.0     2.0   
77        3.0                  3                  2010.0    12.0   
78        3.0                  3                  2011.0    12.0   
79        3.0                  3                  2012.0    12.0   
80        3.0                  3                  2013.0    12.0   
81        3.0                  3                  2014.0    12.0   
82        3.0                  3                  2015.0     5.0   
83        3.0                  3                  2005.0     8.0   
84        3.0                  3                  2006.0    12.0   
85        3.0                  3                  2007.0    12.0   
86        3.0                  3                  2008.0    12.0   
87        3.0                  3                  2009.0    12.0   
88        3.0                  3                  2010.0    12.0   
89        3.0                  3                  2011.0    12.0   
90        3.0                  3                  2012.0     2.0   
91        3.0                  3                  2013.0     1.0   
92        3.0                  3                  2014.0    12.0   
93        3.0                  3                  2015.0     5.0   
94        3.0                  3                  2008.0     3.0   
95        3.0                  3                  2009.0    12.0   
96        3.0                  3                  2010.0    

In [6]:
df_bart.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190571 entries, 0 to 190570
Data columns (total 25 columns):
AGENCY_ID                  190569 non-null float64
PRIMARY_AGENCY_ID          190571 non-null int64
STAT_PROFILE_DATE_YEAR     190570 non-null float64
MONTHS                     190570 non-null float64
PRD_INCRD_LOSSES_AMT       190570 non-null float64
RETENTION_RATIO            190570 non-null float64
LOSS_RATIO                 190570 non-null float64
LOSS_RATIO_3YR             190570 non-null float64
GROWTH_RATE_3YR            190570 non-null float64
CL_BOUND_CT_MDS            190570 non-null float64
CL_QUO_CT_MDS              190570 non-null float64
CL_BOUND_CT_SBZ            190570 non-null float64
CL_QUO_CT_SBZ              190570 non-null float64
CL_BOUND_CT_eQT            190570 non-null float64
CL_QUO_CT_eQT              190570 non-null float64
PL_BOUND_CT_ELINKS         190570 non-null float64
PL_QUO_CT_ELINKS           190570 non-null float64
PL_BOUND_CT_PLRANK      

In [7]:
df_filter = df_bart[(df_bart["MONTHS"] == 12)]
df_filter.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 141221 entries, 1 to 190569
Data columns (total 25 columns):
AGENCY_ID                  141220 non-null float64
PRIMARY_AGENCY_ID          141221 non-null int64
STAT_PROFILE_DATE_YEAR     141221 non-null float64
MONTHS                     141221 non-null float64
PRD_INCRD_LOSSES_AMT       141221 non-null float64
RETENTION_RATIO            141221 non-null float64
LOSS_RATIO                 141221 non-null float64
LOSS_RATIO_3YR             141221 non-null float64
GROWTH_RATE_3YR            141221 non-null float64
CL_BOUND_CT_MDS            141221 non-null float64
CL_QUO_CT_MDS              141221 non-null float64
CL_BOUND_CT_SBZ            141221 non-null float64
CL_QUO_CT_SBZ              141221 non-null float64
CL_BOUND_CT_eQT            141221 non-null float64
CL_QUO_CT_eQT              141221 non-null float64
PL_BOUND_CT_ELINKS         141221 non-null float64
PL_QUO_CT_ELINKS           141221 non-null float64
PL_BOUND_CT_PLRANK      

Removing 50000 rows shouldn't be much of an issue within a dataset like this. We still have plenty to work with.

## CL_BOUNDS_CT_MDS to PL_QUO_CT_TRANSACTNOW
#### Column's description
*
CL_BOUND_CT_MDS – (contains missing values) number of bound policies quoted through a MDS (probably a data recording error, should be DSM) in the current year to date, that is the first six months of 2015, in commercial lines • CL_QUO_CT_MDS – (contains missing values) number of quoted policies through a MDS (probably a data recording error, should be DSM) in the current year to date, that is the first six months of 2015, in commercial lines • CL_BOUND_CT_SBZ – (contains missing values) number of bound policies quoted through a SBZ in the current year to date, that is the first six months of 2015, in commercial lines • CL_QUO_CT_SBZ – (contains missing values) number of quoted policies through a SBZ in the current year to date, that is the first six months of 2015, in commercial lines • CL_BOUND_CT_eQT – (contains missing values) number of bound policies quoted through an eQT in the current year to date, that is the first six months of 2015, in commercial lines • CL_QUO_CT_eQT – (contains missing values) number of quoted policies though an eQT in the current year to date, that is the first six months of 2015, in commercial lines • PL_BOUND_CT_ELINKS – (contains missing values) number of bound policies quoted through ELINKS since September 2013 in personal lines • PL_QUO_CT_ELINKS – (contains missing values) number of quoted policies though ELINKS since September 2013 in personal lines • PL_BOUND_CT_PLRANK – (contains missing values) number of bound policies quoted through PLRANK since September 2013 in personal lines • PL_QUO_CT_PLRANK – (contains missing values) number of quoted policies though PLRANK since September 2013 in personal lines • PL_BOUND_CT_eQTte – (contains missing values) number of bound policies quoted through eQTte since September 2013 in personal lines • PL_QUO_CT_eQTte – (contains missing values) number of quoted policies though eQTte since September 2013 in personal lines • PL_BOUND_CT_APPLIED – (contains missing values) number of bound policies quoted through APPLIED since September 2013 in personal lines • PL_QUO_CT_APPLIED – (contains missing values) number of quoted policies though APPLIED since September 2013 in personal lines • PL_BOUND_CT_TRANSACTNOW – (contains missing values) number of bound policies quoted through TRANSACTNOW since September 2013 in personal lines • PL_QUO_CT_TRANSACTNOW – (contains missing values) number of quoted policies though TRANSACTNOW since September 2013 in personal lines
*

These tables describe how much (bound) policies were quoted through another *unit*.
Since these go per 2, we can seperate them from the rest and look at which unit is doing the "best". There is missing data within these columns, all these 9999 values can be replaced with 0 without making much of an inpact.

If a product is in the PL we can replace all 9999 with a 0 in the CL Columns, this value isn't used. Same goes from the CL in the PL columns.
## Loss ratio and Loss Ratio 3YR
These values seem to be the loss percentages that the company in questions have gotten. This data could be worked into a label by taking all positive numbers (losses) as 0 and negative numbers (gains) as 1. There are also 0s in there, this could be that the company was on the breakeven point or that this value is missing. To use this column as a label we will have to delete all the missing data in it. Filtering these values would leave us at 55946 rows out of 190571, a big loss of data.

We could fill in the blank spots with an average but there is not a lot of data.
## Growth rate 3YR

#### Columns Description
*
GROWTH_RATE_3YR – (computed & contains missing values) computed by agency by line of business by year for the three year period ending in that year; measures the average growth in written premium for that agency in that line of business; only computes results for agencies that have data for the entire range of years; since the measure is over three years of growth, there needs to be a base year to be used as a standard so four years of data are needed; in order to include as many results as possible, the PREV_WRTN_PREM_AMT column is used if it exists in the first year of data available, so that it can be used as a base year, otherwise the WRTN_PREM_AMT is the only column used
*


**There needs to be 4 FULL years of data for this number to be computed**. The increase in the amount of premiums written within those years. Could be usefull for indication and the amount of missing data is manageable. 108941 out of 190751 records have data within them. A very usable column.

Since there is a lot of data within this column I can recommend filling in the missing data with averages IF the company or products excists at least 4 years.
## Retention Ratio
#### Column's description
*
RETENTION_RATIO – (computed & contains missing values) computed for each row in the data as RETENTION_POLY_QTY / PREV_POLY_INFORCE_QTY; therefore it’s a granular measure of the retention for that agency writing that particular product in that particular state from the previous year
*

Too many missing values, excel bugs out and doesn't recognise it but I think at least 50-75% is missing. Meaning of the table is awkward and weird. Not useful to us
## PRD_INCRD_LOSSES_AMT
#### Column's description
*
PRD_INCRD_LOSSES_AMT – losses
*

Losses. 14625 out of 190571 records left when filtered. No information on what this table means. Unusable.


# Conclusion
**Growth rate** will be a very important column to look at seeing its size and meaningful data within it.

**Loss ratio** is a nice backup to have but has a lot of missing values.

**PRD_INCRD_LOSSES_AMT** has a lot of values that could be used but are currently meaningless.

**Retention Ratio** weird data with lots of missing values without much of a meaning.

**Garbage data at the end** usefull when looking at outside sources but not that meaningful on its own